allennlp/training/trainer.py

```diff
@@ -775,14 +763,15 @@ def from_params(  # type: ignore
         lr_scheduler_params = params.pop("learning_rate_scheduler", None)
         momentum_scheduler_params = params.pop("momentum_scheduler", None)
 
+        check_for_gpu(cuda_device)
         if isinstance(cuda_device, list):
-            model_device = cuda_device[0]
-        else:
-            model_device = cuda_device
-        if model_device >= 0:
+            raise ConfigurationError(
```

Comment:
I would like to move this check inside `parse_cuda_devices`, as the two places we pass cuda devices are strictly typed - `cuda_device` inside the trainer config must be an `int` and `distributed_cuda_devices` at the top level must be a `List[int]`. However, we should probably still raise a nice warning for people who are converting their configs. Thoughts?

Reply:
I'm definitely in favor of raising a nice warning to assist in changing the configs. To that end, could we add a snippet in the exception? Something like:

```
"""In allennlp 1.0, the Trainer cannot be passed multiple cuda devices. Instead, use the faster Distributed Data Parallel. For instance, if you previously had config like:
{
  "trainer": {
    "cuda_device": [0, 1, 2, 3],
    "num_epochs": 20,
    ...
  }
}

simply change it to:

{
  "distributed": {
    "cuda_devices": [0, 1, 2, 3],
  },
  "trainer": {
    "num_epochs": 20,
    ...
  }
}
```

Reply:
As for where to put the warning, it seems that `parse_cuda_devices` is outdated now given the strict typing you mention. Maybe make a `parse_cuda_device` that gives an error if you pass it a list and similarly a `parse_cuda_devices` that mandates it has a list?

Reply:
Added this inside `parse_cuda_device`, which now strictly returns an `int`. We ended up not needing the version which parses multiple of them, as we only do it in one place.